File: java/bob/src/main/java/Bob.java
import java.util.Arrays;
import java.util.function.Predicate;

public class Bob {

    enum Type {
        QUESTION(isSilent, "Fine. Be that way!"),
        YELL(isYelling, "Whoa, chill out!"),
        NOTHING(isQuestion, "Sure."),
        WHATEVER(sentence -> true, "Whatever.");

        Predicate<String> predicate;
        String sentence;

        Type(Predicate<String> predicate, String sentence) {
            this.predicate = predicate;
            this.sentence = sentence;
        }
    }

    private static Predicate<String> isSilent = sentence -> sentence.trim().isEmpty();

    private static Predicate<String> isYelling = sentence -> sentence.matches(".*[a-zA-Z]+.*")
            && sentence.toUpperCase().equals(sentence);

    private static Predicate<String> isQuestion = sentence -> sentence.endsWith("?");

    public String hey(String sentence) {


        return Arrays.stream(Type.values())
                .filter(e -> e.predicate.test(sentence))
                .findFirst()
                .get().sentence;
    }

}

File: go/bob/bob.go
// Package bob : Contains function to talk to Bob
package bob

import (
	"strings"
	"unicode"
)

// Hey is the only way to comunicate with Bob
func Hey(remark string) string {
	s := strings.TrimSpace(remark)
	switch {
	case isQuestion(s) && isYelling(s):
		return "Calm down, I know what I'm doing!"
	case isQuestion(s):
		return "Sure."
	case isYelling(s):
		return "Whoa, chill out!"
	case isSilence(s):
		return "Fine. Be that way!"
	default:
		return "Whatever."
	}
}

func isYelling(s string) bool {
	hasLetters := strings.IndexFunc(s, unicode.IsLetter) >= 0
	isUpper := strings.ToUpper(s) == s

	return hasLetters && isUpper
}

func isQuestion(s string) bool {
	return strings.HasSuffix(s, "?")
}

func isSilence(s string) bool {
	return len(s) == 0	
}
